src/System.Management.Automation/utils/tracing/SysLogProvider.cs

```diff
@@ -0,0 +1,504 @@
+    /// and resource string reference.
+    /// </summary>
+    /// <remarks>
+    /// This component logs ETW trace events to syslog.
+    /// The log entries use the following common format
+    ///     (commitId:threadId:channelid) [context] payload
+    /// Where:
+    ///     commitId: A hash code of the full git commit id string
+    ///     threadid: The thread identifier of calling code.
+    ///     channelid: The identifier for the output channel. See PSChannel for values.
+    ///     context: Dependent on the type of log entry
+    ///     payload: Dependent on the type of log entry.
+    /// Note:
+    ///     commitId, threadId, and eventId are logged as HEX without a leading
+    ///     '0x'.
+    ///
+    /// 4 types of log entries are produced.
+    /// NOTE: Where constant string are logged, the template places the string in
+    /// double quotes. For example, the GitCommitId log entry uses "GitCommitId"
+    /// for the context value.
+    ///
+    /// Note that the examples illustrate the output from SysLogProvider.Log,
+    /// Data automatically prepended by syslog, such as timestamp, hostname, ident,
+    /// and processid are not shown
+    ///
+    /// GitCommitId
+    ///   This is the first log entry for a session. It provides a correlation
+    ///   between the full git commit id string and a hash code used for subsequent
+    ///   log entries.
+    ///   Context: "GitCommitId"
+    ///   Payload: string "Hash:" hashcode
+    ///    where string is the full git commit id string and hashcode is the integer
+    ///    hash code for the string in HEX.
+    ///   Example: (19E1025:3:10) [GitCommitId] {eaa9aec8-5b5f-4f68-b045-5e9ae3cfc0b5} Hash:193E1025
+    ///
+    /// Transfer
+    ///   A log entry to record a transfer event.
+    ///   Context: "Transfer"
+    ///   The playload is two, space separated string guids, the first being the
+    ///   parent activityid followed by the new activityid
+    ///   Example: (19E1025:3:10) [Transfer] {de168a71-6bb9-47e4-8712-bc02506d98be} {ab0077f6-c042-4728-be76-f688cfb1b054}
+    ///
+    /// Activity
+    ///   A log entry for when activity is set.
+    ///   Context: "Activity"
+    ///   Payload: The string guid of the activity id.
+    ///   Example: (19E1025:3:10) [Activity] {ab0077f6-c042-4728-be76-f688cfb1b054}
+    ///
+    ///  Event
+    ///   Application logging (Events)
+    ///   Context: EventId:taskname.opcodename.levelname
+    ///   Payload: The event's message text formatted with arguments from the caller.
+    ///   Example: (19E1025:3:10) [Perftrack_ConsoleStartupStart:PowershellConsoleStartup.WinStart.Informational] PowerShell console is starting up
+    /// </remarks>
+    internal class SysLogProvider
+    {
+        // The full git commit id string
+        static readonly string _gitCommitId;
+        // The hash code for the git commit id.
+        static readonly int _commitId;
```

Comment:
FYI - storing the hash code isn't necessary, it's highly optimized - stored in the string so it's only computed once. It's probably slower to store it yourself (not verified though).

Reply:
Where are you seeing this implementation? When I review GetHashCode in the coreclr code base, I don't see any caching of the hashcode.

Reply:
It's not exposed via managed code - it's a low level implementation detail because `string` needs to be highly optimized.

Reply:
Does it also optimize away the function call?

Reply:
After reading about String.GetHashCode, I'm going to leave the implementation as is.  There's a good argument for generating a hash using an algorithm that is repeatable and less likely to collide. As it stands, it's not hard to get two hash codes for the same commit id.

I'll add a comment to that effect.

Reply:
Why do we need the hash code for git commit id? Is it used as some sort of an identification for a logging?

Reply:
The hash code is used in-lieu of logging the long commit id with every log message.  Instead, I log an initial entry containing the full commit id and the associated hash. Subsequent entries contain on the hash.  SysLogProvider has a comment block at the beginning that illustrates its usage.

Reply:
I suppose you can ignore my comment about caching - it's not clear to me that caching is really happening.

If you feel like digging into details, you could start [here](http://mattwarren.org/2016/05/31/Strings-and-the-CLR-a-Special-Relationship/#comment-2727920301).